src/bevy_compat/keyboard.rs

```diff
@@ -0,0 +1,902 @@
+
+impl Default for Modifiers {
+    fn default() -> Self {
+        Self(KeyModifiers::empty())
+    }
+}
+
+// I wish KeyboardInput had the Hash derive.
+//
+// TODO: Drop this if KeyboardInput gets a Hash impl.
+// [PR](https://github.com/bevyengine/bevy/pull/14263)
+#[derive(Deref, DerefMut, PartialEq, Eq)]
+struct KeyInput(KeyboardInput);
+
+impl Hash for KeyInput {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: Hasher,
+    {
+        self.key_code.hash(state);
+        self.logical_key.hash(state);
+        self.state.hash(state);
+        self.window.hash(state);
+    }
+}
+
+/// This is a double-buffer for the last keys pressed.
+///
+/// - `.0` is current frame's last keys pressed.
+/// - `.1` is the next frame's last keys pressed.
+///
+/// During processing we read from `.0` and write to `.1`.
+#[derive(Default)]
+struct LastPress(HashSet<KeyInput>, HashSet<KeyInput>);
+
+impl LastPress {
+    fn swap(&mut self) {
+        std::mem::swap(&mut self.0, &mut self.1);
+    }
+}
+
+/// If the terminal does not support sending key release events, this plugin
+/// tries to emulate them. An event stream might look like this:
+///
+/// `Press A, Press B`
+///
+/// This plugin will emit events:
+///
+/// `Press A, Release A, Press B, (timer elapses) Release B`
+///
+/// The timer is set to one second by default but it can be overwritten.
+#[derive(Resource, Debug, Deref, DerefMut)]
+pub struct ReleaseKey(pub Timer);
+// XXX: This could be structured differently:
+// enum ReleaseKey {
+//     Never,
+//     FrameCount(u8),
+//     Duration(Duration),
+//     Immediate,
+// }
+
+impl Default for ReleaseKey {
+    /// Set the release key timer for 1 second by default.
+    fn default() -> Self {
+        ReleaseKey(Timer::from_seconds(1.0, TimerMode::Once))
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn send_key_events_with_emulation(
+    mut keys: EventReader<KeyEvent>,
+    window: Query<Entity, With<PrimaryWindow>>,
+    mut modifiers: Local<Modifiers>,
+    mut last_pressed: Local<LastPress>,
+    mut keyboard_input: EventWriter<KeyboardInput>,
+    mut timer: ResMut<ReleaseKey>,
+    time: Res<Time>,
+    mut detected: ResMut<Detected>,
+    policy: Res<EmulationPolicy>,
+    backlog: Local<Vec<KeyboardInput>>,
+) {
+    if policy.emulate_what(&detected).is_empty() {
+        send_key_events_no_emulation(keys, window, keyboard_input, backlog);
+        return;
+    }
+    timer.tick(time.delta());
+    if keys.is_empty() && !timer.finished() {
+        return;
+    }
+
+    let bevy_window = window.single();
+    for key_event in keys.read() {
+        if matches!(key_event.code, crossterm::event::KeyCode::Modifier(_)) {
+            detected.0 |= Capability::MODIFIER;
+        }
+
+        if let Some((bevy_event, mods, repeated)) = key_event_to_bevy(key_event, bevy_window) {
+            if bevy_event.state == ButtonState::Released {
+                detected.0 |= Capability::KEY_RELEASE;
+            }
```

Comment:
This can be. If we can put them in different systems, it might be nice to use bevy_states, but I don't know how you feel about adding dependencies.

Reply:
I think this can probably be achieved without the burden of using bevy_states. I guess it might add the ability to inspect what the detection state currently is externally. Are there other plugin examples that use state like you're suggesting?  Otherwise it seems like states won't add much to this over just keeping track of whether the detection has occurred in a resource. Essentially the same as what's already happening in the if statements in this function.

Reply:
It would allow for more idiomatic bevy because you could split the strategies into separate systems that are scheduled to run in specific states (providing more info to bevy's scheduler than conditionals inside the system). However I'm agnostic about whether this is worth adding the whole dependency.

Reply:
How about skipping adding it for now and implementing it as a followup?

I don't really mind adding the dependency, it's the complexity that I'd prefer to avoid. It might be that the states approach is less complex, more useful, and easier to reason about, but a separate PR that just does that change would highlight that (and make it easier to reason about the impact).